WIP: NDData several bug fixes and added functionality (Refactoring it)

MSeifert04 opened the pull request:
These are just some things I have been changing on my local computer to get some more functionality out of NDDataBase and it's subclasses. This is not very complete (I haven't touched NDArithmeticMixin as much as I wanted so most of the tests will break for the latest commits) but I am still working on it.

In the next days I can provide some more information (if wanted) for now this is just to get some feedback if that kind of refactoring would be appreciated or not (there are some API changes in there but I tried to keep them low). Unfortunatly I have forgotten the skip ci on some of the commits but since these are very few I hope that's not too bad.

I hope it is ok to just create a PR for such a large change.

MSeifert04 commented:
Just as a short summary:
- NDData can now be initialized with Masked_Quantities
- NDSlicingMixin now only tries to slice properties if they match certain criterias
- NDUncertainty is completly changed (moved lots of things from StdDevUncertainty to base class)
- StdDevUncertainty can now process correlations and yields correct results if one uncertainty is missing.
- I think I also made it easier to subclass from NDSlicingMixin to allow for slicing of additional properties and to subclass from NDUncertainty.
- I updated the docs inside the files to create more helpful (but since english is not my strong side - maybe more uncomprehensable) documentations.
- I updated most test cases to match my changes.
- API changes mostly affect NDUncertainty (but not StdDevUncertainty).
In the next days my goal is to revisit NDArithmeticMixin to match the form of the new NDSlicingMixin and to adapt to the new NDUncertainty API.
Any Feedback is welcome and if this refactor is unwanted I have no problems withdrawing it again.

mwcraig commented:
@MSeifert04 -- a large PR is appropriate for this, I think. It is easier to see the changes all in one place. I'll do a quick review right now looking for minor things, then a more thoughtful review Monday. After that I'll probably put out a call on the astropy-dev list for people to come take a look if they are interested.

Just a heads up -- it is typical for a pull request of this size to end up going through a bunch of revisions.

Could you please push a trivial change (could be whitespace or whatever) to your branch? That will trigger the test suite. [I know there will be some fails but it will be useful to see where.]

For the future you don't need to put all of the `[skip ci]`. When you open a PR with a lot of commits the tests only run against the last commit, not individually against each commit.

mwcraig commented:
@wkerzendorf -- would be good to get a second set of eyes on this.

@MSeifert04 -- one other thought...this may end up getting split into smaller PRs if some things affect 1.0.x but others are enhancements. Like I said above, though, useful for now to see it all in one place.

mwcraig commented on astropy/nddata/mixins/ndslicing.py:
```diff
@@ -11,29 +18,224 @@ class NDSlicingMixin(object):
+       - if the ``data`` is not a `~numpy.ndarray` or should not be sliceable
+         you may be better off not using this Mixin.
+       - if you have defined an additional property that needs to be sliced
+         extend the ``_slice`` method. Propably the best way to do this is
```
`Propably` → `Probably`

MSeifert04 replied:
Ok, I fixed this and pushed it to get the tests started.

mwcraig commented on astropy/nddata/nddata.py:
```diff
@@ -37,20 +30,15 @@ class NDData(NDDataBase):
     uncertainty : any type, optional
-        Uncertainty on the data. The uncertainty *must* have a string attribute
-        named ``uncertainty_type``, but there is otherwise no restriction.
+        Uncertainty on the data. The uncertainty *should* have a string
```
I'm think we'll need to stick with `*must*` -- this point got a ton of discussion in APE 7....

MSeifert04 replied:
I might be on your side but that's mostly equivalent on enforcing the uncertainty being a `NDUncertainty`. Since having an attribute requires a customized class and having uncertainty propagation requires that this class has an interface that matches the workflow of `ArithmeticMixin`. Since I'm not nearly finished with the arithmetic and propagation (propagation should work, but there are some unresolved issues regarding unit comparison I need to investigate) this requires the end-user or package that uses the `NDData` to emulate `NDUncertainty` or use it. 

But in some cases one might just consider saving some uncertainty sources there (maybe a dict) or the class just assumes standard-deviation or variance and works with that. I think issueing a warning when there is no such attribute should be enough. That forces someone who works with `NDData` to either endure the `log.Info`s or use `NDUncertainty` (or something like it). Currently if one wants to use some simpler form like just a  numpy array or dict one needs to subclass NDData and overwrite the uncertainty setter. With the `should` it is inconvieniant to use something else but one does not need any programming skills to use it.

I can revert the behaviour back to `must` but I really tried not to enforce anything but to suggest it instead.

MSeifert04 replied:
@mwcraig I just realized that there is already a PR which implements an ``UnknownUncertainty`` class (#3714) if you really do not like the *should* I could implement an adapted version of ``UnknownUncertainty``. I could alter the ``getter`` so that anyone using an unknown uncertainty is 'believing' ``NDData`` has saved their ``value`` but in truth it is something that subclasses from ``NDUncertainty`` with the corresponding properties and methods (but altered in a way that it cannot be propagated with anything not even with itself). That would actually be only be a minor change.

mwcraig commented on astropy/nddata/nddata.py:
```diff
@@ -180,10 +178,25 @@ def __repr__(self):
+        """
+        `~numpy.ndarray`: the data
+
+        Even though it is only enforced that the `data` is `~numpy.ndarray`
```
Do you mean that the data is only required to be `numpy.ndarray`-like?

MSeifert04 replied:
I think that may be a residual of when I wanted to move some of the `init` logic to specific setters. And the text defines how the setter originally worked (*must be an ndarray* because overwriting the data with something else would be ambiguous). Should I delete the whole text after 'Even though ...'  in that docstring?

mwcraig commented on astropy/nddata/nddata.py:
```diff
@@ -180,10 +178,25 @@ def __repr__(self):
+
+        Even though it is only enforced that the `data` is `~numpy.ndarray`
+        which does not contain objects. For use with `NDArithmeticMixin` it is
+        assumed that the data array only contains numbers.
```
I'd move the the sentence that starts `For use with` to the `NDArithmeticMixin`.

MSeifert04 replied:
Good that you pointed that out. `NDData` only prevents the `data.dtype` to be `O` (objects). I think the most common case is that the `data` is only numbers but what if they are strings? Should I edit the `init` to prevent that? This documentation should simply highlight that `non numbers` cannot be used with arithmetic operations. Nevertheless I will move it to the class docstring of the arithmeticMixin.

mwcraig commented on astropy/nddata/mixins/ndslicing.py:
```diff
@@ -11,29 +18,224 @@ class NDSlicingMixin(object):
+    to rewrite the whole Mixin (see Notes for detailed suggestions on how
+    to do that).
+
+    Notes
```
I wonder if the bulk of the notes here should be added to the astropy documentation. The advice about subclassing is useful, but might be easier to find in the main documentation, with a reference here to those docs. I'm referring to this: https://github.com/astropy/astropy/blob/master/docs/nddata/subclassing.rst

MSeifert04 replied:
That's right. I totally ignored altering the main documentation. I think after I finished with `NDArithmeticMixin` I should consider which part of the code documenation should be moved to the online documentation.

mwcraig commented on astropy/nddata/mixins/ndslicing.py:
```diff
@@ -11,29 +18,224 @@ class NDSlicingMixin(object):
+            else:
+                log.info("Uncertainty has not the same shape as data.")
+        elif isinstance(self._uncertainty, NDUncertainty):
+            if self._uncertainty.array.shape == self.data.shape:
```
Why does the shape need to be checked here? If it is going to be checked shouldn't that be done in a setter? Or is the worry that slicing might have somehow changed the shape of one and not the other?

MSeifert04 replied:
Currently `NDUncertainty` defined no setter for the `array` (which contains the uncertainty values) so subclasses might not implement shape checks in the setter or `init`. So there were two possibilities:
1.) enforcing the uncertainty to have the same shape as the data, which is a tremendeous effort because there might be use-cases were we only need the uncertainty to be broadcastable to the data's shape. 
2.) which brings me to the second point: If we don't enforce the shapes to match we would need to implement how slicing affects something that is only broadcastable to the data shape, this is not so complicated but the most common case is, when the uncertainty is either a scalar or a ndarray with the same shape as the data. In either case the current slicing works (ok it issues a warning if it is a scalar but it does not try to slice it).

If someone really want's to slice more complicated uncertainties (which are only broadcastable to data shape or as you pointed out with the mask some lazy-evaluation function) the subclass has to override the `_slice_uncertainty` method.

Or is the requirement that the uncertainty has the same shape as the data something that should be enforced?

mwcraig commented:
@MSeifert04 -- I'll have more Monday (and maybe even later today if the day goes well), but a few broad comments now:

+ I like separating the tests into separate files, and the fact that you've added many!
+ The more extensive documentation about subclassing is very helpful. As mentioned above, it might make sense to collect it together in https://github.com/astropy/astropy/blob/master/docs/nddata/subclassing.rst
+ The slicing changes are currently to restrictive, I think. Uncertainty or mask, for example, could be functions (something that never even occurred to me until someone pointed it out to me when doing the big NDData reorg), or an array-like object whose entries are lazy evaluated as needed. What is the underlying reason for proposing restricting them?

Very excited to see this extensive work, and more comments Monday!

MSeifert04 commented:
@mwcraig Thanks for the comments and the heads up. I will take a closer look at your suggestions later. For now I started the tests and quickly wrote some answers to your comments. Later today or tomorrow I hope I get the `NDArithmeticMixin` and `StdDevUncertainty` working properly and then I will have some more time considering where to enforce/restrict something. I changed several things then tried to apply slicing/aritmetics and found that I have to change them again or revert the changes (which is a pain while writing documentations and keeping track of when I changed what and why and where or if to mention them). For now I tried to apply the least bit of restrictions everywhere while allow everything to work (optional with some warnings) as one would expect.

Thanks again for your kind words, because at first I was not sure if I should really pull this branch. Mostly because some changes are just because `I thought I could use that` (e.g. splitting the `__getitem__` into several methods for each property) or `I don't need that restriction` (e.g. Uncertainty has to match parents shape was really annoying in my first attempts to calculate the resulting uncertainty if one uncertainty is missing since just setting it to a scalar zero did not work).

MSeifert04 commented:
Ok, I'm finally satisfied with the first version of `NDArithmeticMixin`, there are some new features and some bugfixes in there but I hadn't the time (yet) to finish the tests. There is still some subclassing documentation in there too and I hadn't much time to adress the open comments of @mwcraig but I will get to that tomorrow. Just as a short summary:
- The bug with one uncertainty missing kills uncertainty propagation is fixed in there
- Masks are now checked (if both are set) that they are `booleans` or arrays of these. That may be too restricted (as with slicing) but at least for now it's behaving similar in both mixins (if I change it again I can change both, but the mixins are not inconsistent right now)
- I skipped the `unit` test before the data is operated because my timings said there is not a real time difference between a failed 1 element operation and a failed 1000x1000 element operation.
- I added a premiliminary meta-handling method. But I'm not very happy with the documentation of this one. The functionality is propably a bit unstable too because the requirements are quite hard.
- I changed the uncertainty propagation to meet the new entry point in `NDUncertainty` which is the `propagate` method and not the specific `propagate_*` method. (but I haven't got around to really test the new StdDevUncertainty especially the ``correlation`` and ``uncertainty has a different unit than the data`` cases)
- wcs comparison is now optional and even if turned on only issues warnings not exceptions. (At least I had some problems using any kind of WCS because == does not work in my examples)
- I changed the API of `NDArithmeticMixin.add` (and similar) quite a lot, there are some new parameters possible to regulate the operation.
- I added classmethods to allow for a - b where b is some class with ArithmeticMixin and a is just a scalar, numpy array, quantity or ... these methods are called `NDArithmeticMixin.addition` and take two operand parameters and then refer to `add`. The special usage here is that the first operand is wrapped in a class instance of the instance that calls the `addition` so this also allows for easy arithmetic operations between different subclasses of `NDData`. (I don't know if a classmethod is a good idea, most people do not recommend it but I thought it would be a great place for the my first use of classmethods ever)
- In the spirit of making arithmetics possible with other types I also converted the second operand (but in the `arithmetic` method) to the same class as the first operand.

The internal flow currently is (i.e. for addition):
[`addition` ->] `add` -> `arithmetics` (only collects the results of) -> `_arithmetics_*`.
Especially the case of NDUncertainty should be mentioned since I changed the behaviour there too
`propagate` -> `_convert_uncertainty` (checks if the other uncertainty can be used for propagation) -> `_propagate_add` (calculates)

These method seperations were because I had some problems while subclassing to adapt to new properties or other restrictions. This may or may not be sensable in every case but it allows that any subclass just keeps the parts it wants and overrides or extends the methods that do not suit it. Since this is one of my first medium sized projects I'm open for suggestions if that flow should be modifyed or reverted to the one method state.
@mwcraig I will wait with the writing of the `.rst` documentations until some basic issues you mentioned, especially these which adress the enforcing of properties, are adressed and resolved.

embray commented:
At some point I'll probably be asking you to rebase this PR, but let's get it all working first :)  Thanks for all your work on this!

wkerzendorf commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -8,10 +8,16 @@
 import numpy as np
 
-from ... units import dimensionless_unscaled, UnitsError
+from ...units import dimensionless_unscaled, Quantity
```
a few style guides suggest using absolute imports: `from astropy.units import dimensionless_unscaled`, etc. @embray might know more about these style guides.

embray replied:
Astropy has always used relative imports.  Guido doesn't like them for some reason but other people disagree with him on that.

MSeifert04 replied:
@wkerzendorf @embray I'll stick with the relative imports for now. If it shows that there is a downside to it or isn't considered good coding style I'll change them.

wkerzendorf commented:
@MSeifert04 thanks for the work. I would feel that it is much easier if you split your work up into separate PRs. I think that will also ensure that things get merged much faster. Is it easily possible to pull out some of the classes you worked on that have changes that are not controversial? @embray @mwcraig

MSeifert04 commented:
@wkerzendorf :I think it is not so easy to split up the changes into seperate PRs without losing at least some of the changes. Most of the changes are based on a slightly less restrictive ``NDDataBase`` (which is controversial as @mwcraig pointed out) and on a lot of changes in ``NDUncertainty`` (there are no comments on how I changed it so that may be controversial). There is no single class that could be merged without having at least one of the other changed classes. I could alter these classes slightly but I fear that will result in a lots of rebasing actions later on. For now the only controversial point seems to be the lessened restrictions on the uncertainty property. The other comments are rather technical, or am I misinterpreting @mwcraig comments? But since I am relying on you to point me to controversial points that I missed in all the seperate discussions on APE7 (which are in the dev mailing list, propably also on the normal mailing list, on APE7, on some of the APE7 PRs and I may have missed some important points or forgot about it) so if you wish that I split this up I will gladly do it.

MSeifert04 commented:
I have a problems about the test failures, I will now skip the failing doctest in ``NDSlicingMixin`` with ``__doctest_skip__ = ['NDSlicingMixin']`` but is that enough? On my installation the doctest did not fail before so I cannot reproduce the failure and am not sure if adding this one line is enough to stop the doctest from failing.

The other question is about the ``Documentation build`` with sphinx, I altered most of the problems my local sphinx build complained about but there are a few problems remaining that I cannot figure out why they do not work. Could someone help me there? Or should I ignore these documentation failures for now because lot's of the class documentation should be moved to the main documentation ``.rst`` files?

MSeifert04 commented:
Sorry for writing again (and that I run the CI again) but there were some tests that failed that did not fail on my local installation (even in the python2 environment). Unfortunatly I still did not came around to fully test the new ``NDArithmeticMixin``, especially the new meta handling method is not tested with py.test since I'm struggling with the question of how to make valid - real - tests.

MSeifert04 commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -24,211 +30,585 @@ class NDArithmeticMixin(object):
-    if isinstance(_arithmetic.__doc__, string_types):
-        subtract.__doc__ = _arithmetic.__doc__.format(name="Subtract",
-                                                      operator="-")
+        There are two ways of altering the resulting meta, one is by
```
I'm actually not sure what corner cases might be out there. Currently supported is:
- setting meta keywords to a fixed value and
- ``multiplying`` (or add, ...) meta keywords with the other meta keyword and
- ``multiplying`` (or add, ...) the meta keyword with the data (if it is a scalar) of the other operand. 
Are there other use-cases? These were all I could think of.

mwcraig commented:
@wkerzendorf @MSeifert04 -- this may well end up getting refactored, especially some of the bug fixes that affect 1.0.x. I'm anticipating that I'll need to help @MSeifert04 with that but for the moment this gets all of his changes in one place.

mwcraig commented:
> I'm very sorry for putting so much traffic on the Tests.

Don't worry at all about this -- you are using it appropriately.

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -24,211 +30,585 @@ class NDArithmeticMixin(object):
+            informations). If they do not match a warning is issues.
 
         Returns
         -------
```
I think here we really do want to return an instance of the class -- the result of adding, e.g. two CCD data arrays, should be another one of those arrays.

Ah, I see, that is done lower done.

MSeifert04 replied:
I moved the docs from ``arithmetics`` to ``add`` (and similar) and renamed the arithmetics into ``_arithmetics`` so the rather technical details about the returned ``kwargs`` and similar will not be important for normal users.

mwcraig commented:
@MSeifert04 -- have read through this and will give a fresh look and more comments in the morning. In the short run let's try to get the tests passing (I know it is down to one doctest now) before additional changes.

Are the API changes you made to `add`, etc., primarily additional optional arguments? It looks like that is the case...but if not, please speak up.

MSeifert04 commented:
@mwcraig Thanks for the time looking into it and your comments.

Regarding the tests: I think I found all doctests that failed (ok, I forgot one) and the one remaining sphinx build warning.

Regarding the API changes: ``NDArithmeticMixin.add()`` (and similar): The parameter ``propagate_uncertainties`` is now called ``handle_uncertainties`` which allows for ``None``, ``False``, and ``True``. ``None`` is new and allows to force the result to have no uncertainty. All the other parameters are new and optional. The only other class with API changes that affect backwards compatibility is ``NDUncertainty`` (but not ``StdDevUncertainty``).

embray commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -24,211 +30,588 @@ class NDArithmeticMixin(object):
+            account and combined (by a bitwise ``or`` operation if the
+            :meth:`NDArithmeticMixin._arithmetic_mask` is not overriden).
+
+        handle_mata: `bool` or ``None``
```
`handle_mata`?

MSeifert04 replied:
Renamed it, I'm currently working on the tests of NDArithmeticMixin and when they are done I will push the fix.

embray commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -24,211 +30,588 @@ class NDArithmeticMixin(object):
+            ``handle_uncertainties`` is *not* ``True`` this will be ignored. A
+            value of ``0`` means uncorrelated (which is also the default)
+
+        meta_kwds_operate: ``None`` or `list`
```
I think we're probably going to want to rename or simplify some of these arguments. I haven't had a close look yet though.

MSeifert04 replied:
I'm not very satisfied with the documentation and parameter names too. I already updated at least some of the documentation (there were some inconsistencies about documentation of default values and more) so I hope this comment will not vanish.

embray commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -24,211 +30,588 @@ class NDArithmeticMixin(object):
+        kwargs: `dict`
+            kwargs to create a new instance of the same class as the first
+            operand was. The calling method or function is responsible for
+            creating the instance
```
We'll want to reword this.

MSeifert04 replied:
Any suggestions?

embray commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -24,211 +30,588 @@ class NDArithmeticMixin(object):
+        Parameters
+        ----------
+        operation: `str`
+            see `NDArithmeticMixin.arithmetic` parameter description.
```
Instead of a string this should just take a function as an argument directly.

MSeifert04 replied:
That was the case before I worked on this class and that's a bit tricky since one may call other functions somewhere (like if one wants to process the uncertainties) but I could change that back.

embray replied:
Could you clarify that a bit?  I'm not sure I follow.

MSeifert04 replied:
The function works well for the ``data`` and ``meta`` but for ``uncertainty`` I have to do something like
```
if operator_func is np.add:
    propagate_func = self.propagate_add
elif ...
```
The old version of ``NDArithmetic`` kept ``operation`` (the string describing the operation), ``operator`` (the function that is called for operating on the data) and the ``propagate_func`` (the function that is called to propagate the uncertainties). Since we only need one such parameter because we can deduce the others from that one I randomly chose the ``operation``-string. But I can change that, makes the code a bit shorter for the operation on ``data`` and ``meta``.

embray replied:
What's wrong with that?  It's actually faster.  Could also use a dict, for that matter.  ufuncs (and normal functions) can be used as dict keys.  A getattr could also be used.

This also makes me wonder more about the possibility of using an Operation class, with its propagation method built-in.  That would also open the possibility of adding other operators.

MSeifert04 replied:
@embray I thought about defining it in some container like a ``dict`` but since a major headache for my personal use for the ``NDArithmeticMixin`` is to actually alter the bit which need alteration. Such as extending the arithmetic function to also allow for handling other properties like ``flags``. Currently the ``NDArithmeticMixin`` allows me to use a subclass that defines
```
def arithmetics(operation, **kwargs):
   result, kwargs = super().arithmetics(**kwargs)
   kwargs['flags'] = self._arithmetics_flags(operation)

def _arithmetics_flags(self, operation):
   return 'New Flags'
```
without altering anything else and interpreting the ``operation`` inside this new method like I need it. Using ``dict``s actually would also work but each ``_arithmetics_*`` function would have to setup their own ``dict`` and thats almost like having the ``if... elif``, isn't it?

Regarding the ``Operator`` ... see your other comment in a few minutes.

mwcraig replied:
Regarding `string` vs `function`: the current `NDArithmeticMixin` uses the string `name` in only one place: https://github.com/astropy/astropy/blob/master/astropy/nddata/mixins/ndarithmetic.py#L176

The string name can be obtained from `operation.__name__`, if `operation` is, e.g. `np.add`, so it looks like we could do away with the string name altogether.

embray commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -24,211 +30,588 @@ class NDArithmeticMixin(object):
+            result = operator(self.data * self.unit,
+                              operand.data * operand.unit)
+
+        return result
```
This might all benefit by some kind of operator class like I describe in #3861.  That was specifically in the context of compound models.  But maybe it could have a broader use than that, such as handling on a per-operator basis how to handle uncertainties for that operator, etc.  Not 100% clear if it would help here since I haven't done a deep dive on this yet but it's just a thought.

MSeifert04 replied:
I have read most of the linked PR several times but I don't get it. Do you mean something that actually behaves like https://docs.python.org/3/library/functools.html#functools.singledispatch or am I misinterpreting your suggestion completly? Maybe it's just a problem of me understanding the (rather technical) details.

embray commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -24,211 +30,588 @@ class NDArithmeticMixin(object):
+    @classmethod
+    def division(cls, operand1, operand2, **kwargs):
+        operand1 = cls(operand1)
+        return operand1.divide(operand2, **kwargs)
```
You could combine all of the above 8 methods into 4 [`sharedmethod`](http://docs.astropy.org/en/v1.0.5/api/astropy.utils.decorators.sharedmethod.html#astropy.utils.decorators.sharedmethod)s (without the awkward naming).

MSeifert04 replied:
I can give it a try but all my attempts on using decorators went haywire when it came to the sphinx build since either the ``__doc__`` or the function signature was lost. I will report about the success with ``sharedmethod`` later. Thanks for that tipp.

MSeifert04 replied:
At least in my local sphinx build that really broke some things. The ``@sharedmethod`` kills the documentation of the decorated function and since they behave different (because one needs only the second operand and the classmethod requires both operands) I wouldn't like the classmethod be undecorated. But that could only be my fault. Is there any current use of ``sharedmethod`` with seperate documentation in astropy where I can look how to apply it properly or if that is only a problem on my computer?

embray replied:
That shouldn't be the case.  If it is then that issue needs to be raised separately and I can look into it.  I remember when I first implemented that there was a bug in Sphinx that gave some problems, but astropy-helpers has a workaround to that bug.  Or should, at least.

That said, however it's implemented, I don't like having these separate class methods with subtly different names.  I'd rather see them done away with completely if they can't be merged into the instance methods somehow.

MSeifert04 replied:
I hadn't time yet to check astropy-helpers. Maybe the error originated because I have some local error in there. But could you give me a hint where ``sharedmethod`` is used so I can verify that I did not mess up while setting up the docstrings. Maybe calling to ``__doc__`` of a sharedmethod is somehow intercepted (because of my current ``__doc__`` setters).

As alternative I could rename them from ``addition`` to ``interclass_addition`` that would resolve the name conflict and highlight the proper use of these ``methods`` since they allow to perform operations where the ``return`` is an instance of the class which calls the ``interclass_addition``.

mwcraig replied:
See `CCDData` for an alternative approach to handling operations with scalars: https://github.com/astropy/ccdproc/blob/master/ccdproc/ccddata.py#L238

MSeifert04 replied:
@mwcraig Is this actually on the class_methods or on the scalar-problem? I think I handled the scalar problem on Line 144 in NDArithmeticMixin (I did not get the link to a line in a PR right, sorry) but like the problem with scalars, there is the problem of inverse operations: MaskedArray - CCDData for example. For those cases I wrote the class_methods.

embray commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -24,211 +30,588 @@ class NDArithmeticMixin(object):
+        subtract.__doc__ = doc.format(name="Subtract", operator="-")
+        multiply.__doc__ = doc.format(name="Multiply", operator="*")
+        divide.__doc__ = doc.format(name="Divide", operator="/")
+        del doc
```
Might be good to just make a decorator for this. Not sure if it matters though.

embray commented on astropy/nddata/mixins/ndslicing.py:
```diff
@@ -4,36 +4,241 @@
+        _slice_wcs.__doc__ = doc.format(attr="wcs",
+            allowed="`~numpy.ndarray` with the same shape as the data or a "
+                    "`~astropy.wcs.WCS` object")
+        del doc
```
Okay, yes, definitely make a decorator.

MSeifert04 replied:
No problem, just out of curiosity is there a way to make a decorator keep the function signature and not revert to ``real_function_name(*args, **kwargs)`` because that hides default values and doesn't look so good in the online documentation.

MSeifert04 replied:
Did you mean something like this:
```
def include_docstrings(**kwargs):
    def set_docstring(some_func):
        some_func.__doc__ = """....""".format(**kwargs)
        return some_func
    return set_docstring

class NDArithmeticMixin():
...
    @include_docstrings(name="Add", operator="+",
                    other=":meth:`NDArithmeticMixin.addition`")
    def add(self, operand, **kwargs):
        ...
```
unfortunatly I have not much experience using decorators, do I need to add ``functools.wraps`` or something similar there or will it work this way?

embray commented:
Thanks again @MSeifert04 for all your work on this.  I think it definitely makes sense to have a generic "arithmetic" method with lots of dials to turn (of which the usual operators are just special cases).  Will have to review more deeply later.

astrofrog commented on astropy/nddata/nddata_base.py:
```diff
@@ -14,32 +14,29 @@
         """
-        Mask for the data, following the numpy convention that ``True`` means
-        the data should not be used.
+        Mask for the data, if any.
```
Why change this docstring? It's important to mention the True/False convention.

astrofrog commented on astropy/nddata/nddata_base.py:
```diff
@@ -60,33 +57,13 @@ def wcs(self):
     def meta(self):
         """
-        Metadata, if any, must be dict-like.
+        Meta information, if any.
```
Just to check, why change this too?

MSeifert04 replied:
answer to both your comments: Since ``NDDataBase`` only defines the general interface for the container with meta informations I felt they were wrong there and moved them to ``NDData`` where that is actually enforced or at least recommended (because of the mixins). I can move them back or copy the docstrings of the ``NDData`` class into ``NDDataBase``.

mwcraig replied:
@MSeifert04 -- I think it is useful to have them in both places (one of the rare cases where redundancy is a virtue). I think the reason we don't do any checking of dict-like-ness is that we couldn't come up with a general, but easy, way to do it.

There is no way to check whether the user is using masks in the same sense as numpy masks, so the best we can do is let users of both `NDDataBase` and `NDData` know what they should do.

If the comments you moved to `NDData` were not in there before, please keep them there but also add back to `NDDataBase`.

MSeifert04 commented:
If I now push the new state it will hide some comments even though I haven't adressed them yet. Can I ignore this and push nevertheless or should I adress these issues first and push later?

mwcraig commented:
@MSeifert04 -- why don't you go ahead and address those issues too before pushing. Review on this will  take a while anyway, so might as well keep things more straightforward.

MSeifert04 commented:
@mwcraig You are right but some need some further discussions:

- Your comment about ``must`` on the ``uncertainty_type``. I would be more happy with the ``should`` and the ``astropy.info`` if there is none. But that one is ``outdated`` already since I pushed some change nearby.
- @embray comments about decorators for documentation builds (I got that one working but am not sure if that's the best way to go)
- @embray comments about an ``Operator`` class (which I don't understand completely)
- @embray comment about using ``sharedmethods`` (which I don't get to work with sphinx's documentation build)
 
Also there are some bugs in NDUncertainty and some documentation changes I added since yesterday. Also you proposed that I should get the tests straight before implementing anything else (and since I had done a lot of work and already did many commits locally before the comments came up I'm a bit stuck right now). And most of the comments will stay open but some unresolved ones may be ``outdated`` without being actually outdated but I changed something close by. But for now I will hold back a new push you are right since this PR will take some time I don't need to rush. But I would appreciate some more comments on the open (and closed) issues.

MSeifert04 commented:
I've adressed (at least here in the comments) almost all of the issues. Unfortunatly without any further comments I'm not sure how to commence. Maybe a few questions up front:
- #4242 Is something that I built to use together with this PR, so should I include it in here (and close the other PR) or rather include a temporary version of the decorator in this PR so the tests will continue to work?
- I'm quite happy to have so much feedback an the docstrings but I'm not really good at formulating good and understandable english sentences so with some documentations I know exactly what I want to say (which is not in english) but I cannot write it down.
- I added ``TODO`` tags on all open comments in the code and corrected a lot of documentation/tests (and uncertainty propagation methods), is this enough reason to push it again or is it bad manners? But since @mwcraig s comment 2 days have passed without any significant commenting in here but with a lot of changes locally. 

Maybe saying that I don't want to rush was wrong of me because I really hate having this PR code online where anyone could find my bugs :-)

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -24,211 +30,588 @@ class NDArithmeticMixin(object):
+            If ``None`` the result will have no wcs and no comparison between
+            the wcs of the operands is made. If ``False`` the result will have
+            the wcs of the first operand (or second if the first had ``None``).
+            If ``True`` the resulting mask will be like in the case of
```
`...the resulting mask...`: do you mean `the resulting wcs`?

MSeifert04 replied:
yes.

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -24,211 +30,588 @@ class NDArithmeticMixin(object):
+            ``False`` but the wcs information is checked if it is the same
+            for both operands (in some cases it would be bad to apply any
+            arithmetic operation on datasets which have different wcs
+            informations). If they do not match a warning is issues.
```
Since it is, as you said, bad to combine images with different WCS if the users wants them checked, this should be an exception, not a warning.

MSeifert04 replied:
Since astropy.wcs.WCS does not support ``__eq__`` but only ``is`` checks I thought a warning is more appropriate? The documentation also states that ``wcs`` in ``NDData`` is still experimental (http://docs.astropy.org/en/stable/nddata/nddata.html#wcs)

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -24,211 +30,588 @@ class NDArithmeticMixin(object):
-        propagate_uncertainties : bool
-            Whether to propagate uncertainties following the propagation rules
-            defined by the class used for the ``uncertainty`` attribute.
+        operation: `str`
```
You don't need backticks around the type names for ordinary python types. Single backticks are around some of the arguments to provide a reference to the argument type if the type is from a package like numpy or astropy.

mwcraig replied:
One other sphinx note: double backticks around something generate code-formatted output (like single backticks in github markdown) but single quotes are references.

MSeifert04 replied:
So double backquotes are a Do-Not in source code docstrings? Should I replace `` `str` `` with ``str`` or ``string``? I have looked at the numpy doc guide and looked at astropy doc guide (mostly identical btw) but nowhere found something how/if I should highlight attributes in docstrings.

mwcraig replied:
I couldn't remember what we do so took a peek at the Table docs and couple other places and it looks like current practice is to leave off the double backticks. I don't think it is a big deal to put it in, but you definitely don't want single backticks, I think.

I usually put in the python name for the type when possible, so `str` instead of `string`.

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -24,211 +30,588 @@ class NDArithmeticMixin(object):
+        else:
+            # Make sure these masks are boolean or ndarrays of bools so this
+            # kind of mask operation is possible.
+            if self.mask is not None:
```
I think the checks here are too restrictive -- I agree that the most common use case is going to be numpy arrays, but those need to be `bool` to succeed. For example, two arrays of integers can be bitwise OR'ed. More broadly, there are sensible masks that are not arrays.

More restrictions make sense in `NDDataArray`, which restricts the data to numpy arrays, basically, but `NDArithmeticMixin` should be kept as general as possible.

MSeifert04 replied:
I added the possibilities to ``handle_mask``. If not ``True`` the ``_arithmetic_mask`` is never called (``None`` sets the results mask to None, ``False`` takes the mask of the first operand or if that does not exist the mask of the second operand). And I especially moved mask-arithmetic outside so subclasses that want more than ``None`` and ``False`` only need to override ``_arithmetic_mask`` to implement their wanted behaviour. I'm currently working on the ``.rst`` documentation, to provide examples how to do that.

mwcraig commented on astropy/nddata/mixins/ndarithmetic.py:
```diff
@@ -24,211 +30,588 @@ class NDArithmeticMixin(object):
+        if meta_kwds_operate is not None and len(meta_kwds_operate) > 0:
+
+            # Get the operator function for these operation (use numpy ufuncs)
+            if operation == 'addition':
```
This definitely makes the case using functions as the argument instead of strings!

MSeifert04 replied:
I'll change it even though I still believe the operation-function on the data and the name of the operation on the data might be different. I might be wrong but I thought about what could be saved as data (``numpy.matrix`` and ``pandas.Dataframe``) and the ufunc does not work on them (at least matrix) but the operation name clearly identifies what should be done.

mwcraig commented on astropy/nddata/nddata.py:
```diff
@@ -5,64 +5,55 @@
-        ``mask`` here will causes the mask from the masked array to be
-        ignored.
+    mask : any type, optional
+        Mask for the data.
```
Definitely want the deleted comments about behaving like a numpy mask restored here....

mwcraig commented on astropy/nddata/nddata.py:
```diff
@@ -78,97 +69,122 @@ class NDData(NDDataBase):
+            # and proceed as if that one was given as parameter. Same for the
+            # other parameters.
+            if unit is not None and data.unit is not None:
+                log.info("Overwriting NDData's current "
```
If this is done, don't you need to update the data values also, since the units are changing?

Also, it seems like a message is unnecessary if both units are the same, so maybe check for `and unit is not data.unit`?

MSeifert04 replied:
``overwriting`` means dropping the old one and using the new one. I actually wondered if it makes sense to drop the exception but since all other parameters are overwritten why not the unit? I'll add the ``unit != data.unit`` (``is`` does not work for composite units).

mwcraig commented on astropy/nddata/nddata.py:
```diff
@@ -180,10 +196,21 @@ def __repr__(self):
     @property
     def data(self):
+        """
+        `~numpy.ndarray`: the data.
```
It is definitely good to add a doscstring here -- but the return type will be ndarray-like, not necessarily ndarray.

MSeifert04 replied:
You're right. Changed it.

mwcraig commented on astropy/nddata/nddata.py:
```diff
@@ -180,10 +196,21 @@ def __repr__(self):
+        any type: Mask for the data, if any.
+
+        Using `~numpy.ndarray`-like masks containing ``bools`` is *recommended*
+        if the `NDData` is used for any `~numpy.ma.MaskedArray` operations.
```
I'm not sure this advice is necessary...

mwcraig commented on astropy/nddata/nddata.py:
```diff
@@ -192,12 +219,51 @@ def mask(self, value):
+        any type: wcs for the data, if any.
+
+        Even though nothing is enforced, using `~astropy.wcs.WCS` as ``wcs``
+        attribute is encouraged.
```
I would drop this since right now nothing is enfored.

mwcraig commented on astropy/nddata/nddata.py:
```diff
@@ -192,12 +219,51 @@ def mask(self, value):
+        """
+        any type: Uncertainty in the data, if any.
+
+        Uncertainty *should* have an attribute ``uncertainty_type`` that is
```
This point got a lot of discussion leading up to APE7, the conclusion of which was that uncertainties had to have `uncertainty_type` and that had to be a string. The reasoning was that checking to see if the uncertainty type was `NDUncertainty` was perhaps too restrictive, since one could use some other uncertainty class in principle.

MSeifert04 replied:
That's just a copy of my comments on your previous comment about this:

I share your opinion, but on the other hand that's mostly equivalent to enforcing the uncertainty being a `NDUncertainty`. Since having an attribute requires a customized class and having uncertainty propagation requires that this class has an interface that matches the workflow of `ArithmeticMixin`. Since I'm not nearly finished with the arithmetic and propagation (propagation should work, but there are some unresolved issues regarding unit comparison I need to investigate) this requires the end-user or package that uses the `NDData` to emulate `NDUncertainty` or use it. 

But in some cases one might just consider saving some uncertainty sources there (maybe a dict) or the subclass just assumes standard-deviation or variance and works with that. I think issueing a warning when there is no such attribute should be enough. That forces someone who works with `NDData` to either endure the `log.Info`s or use `NDUncertainty` (or something like it). Currently if one wants to use some simpler form like just a  numpy array or dict one needs to subclass NDData and overwrite the uncertainty setter. With the `should` it is inconvieniant to use something else but one does not need any programming skills to use it.

I can revert the behaviour back to `must` but I really tried not to enforce anything but to suggest it instead.

@mwcraig I just realized that there is already a PR which implements an ``UnknownUncertainty`` class (#3714) if you really do not like the *should* I could implement an adapted version of ``UnknownUncertainty``. I could alter the ``getter`` so that anyone using an unknown uncertainty is 'believing' ``NDData`` has saved their ``value`` but in truth it is something that subclasses from ``NDUncertainty`` with the corresponding properties and methods (but altered in a way that it cannot be propagated with anything not even with itself). That would actually be only be a minor change.

mwcraig commented on astropy/nddata/nddata.py:
```diff
@@ -192,12 +219,51 @@ def mask(self, value):
+
+        Uncertainty *should* have an attribute ``uncertainty_type`` that is
+        a string. `~astropy.nddata.NDUncertainty`-subclasses are recommended,
+        if `~astropy.nddata.NDArithmeticMixin` with uncertainty propagation
```
I wonder if this makes more sense somewhere in the docs for `NDArithmeticMixin`?

mwcraig commented on astropy/nddata/nddata.py:
```diff
@@ -192,12 +219,51 @@ def mask(self, value):
+            # and issue a warning if it has not.
+            if (not hasattr(value, 'uncertainty_type') or
+                    not isinstance(value.uncertainty_type, six.string_types)):
+                log.info('Uncertainty should have attribute uncertainty_type '
```
See above -- I'll also respond to your comments in the main thread of the PR about this.

mwcraig commented on astropy/nddata/nddata_base.py:
```diff
@@ -14,32 +14,29 @@
 class NDDataBase(object):
     """
-    Base metaclass that defines the interface for NDData
+    Base metaclass that defines the interface for NDData.
```
Oops, thanks!

mwcraig commented on astropy/nddata/nddata_base.py:
```diff
@@ -60,33 +57,13 @@ def wcs(self):
-    # In python >= 3.3 it would be easy to decorate one of these (setter or
-    # getter) as abstract but not the other.
-    @property
+    @abstractproperty
```
@embray -- you mentioned elsewhere that there is a way to enforce the requirement that `uncertainty` have an `uncertainty_type` that is a string without providing a concrete implementation in `NDDataBase`...could you point us to some references?

mwcraig commented:
@MSeifert04 -- I added several inline comments today. Thanks again for all of your work on this, and sorry it took a while to give this a thorough look.

> #4242 Is something that I built to use together with this PR, so should I include it in here (and close the other PR) or rather include a temporary version of the decorator in this PR so the tests will continue to work?

Let's keep them separate -- that isn't really nddata specific. The simplest way forward would be to wait until that is merged then rebase this. 

> I'm quite happy to have so much feedback an the docstrings but I'm not really good at formulating good and understandable english sentences so with some documentations I know exactly what I want to say (which is not in english) but I cannot write it down.

No worries, we can get the docstring language cleared up later.

> I added TODO tags on all open comments in the code and corrected a lot of documentation/tests (and uncertainty propagation methods), is this enough reason to push it again or is it bad manners?

Good idea, go ahead and push! It wasn't as much manners I worried about as losing track of changes.

More comments in a moment...

mwcraig commented:
Now that I've had a chance to wrap my head around this I think it would make sense to split this up a bit (we may still be several days from being ready to do that, depending on the other changes you push, so don't start doing it right away :)).

I found it really helpful to have it started all in one place so I could see the full scope of what you are proposing.

What do you think (including places where it might not actually work) of this:

+ One PR for the `NDSlicing` changes and tests, which look like they separate pretty cleanly from everything else.
+ One PR for the updated `NDArithmeticMixin` -- getting that decoupled from the changes in `NDUncertainty` would require some work, but I can help with that.
+ One PR for `NDUncertainty`, although if needed it could be combined with `NDArithmeticMixin`.
+ One PR for the `NDData`/`NDDataBase` changes (which I *think* decouple form the rest?).

That would narrow the scope, and amount of code, in each PR...let's revisit the idea of splitting once your local changes are pushed.

mwcraig commented:
@MSeifert04 -- a heads up about timing: I may have some time Tuesday or Thursday to do a thorough read of the natch of changes, but it might not happen until next Saturday. Wish I could guarantee it will be sooner, but figured better to know what to expect than not :/

MSeifert04 commented:
@mwcraig Thanks for looking into it. I've changed most things you mentioned and those I did not change I marked as ``TODO`` in the files. Actually ``NDArithmeticMixin`` is now polluted with TODOs. But I think I have lost some track about when I changed something since I was holding back from pushing for some days now. I hope that not too much is lost.

Regardings the splitting: I think splitting can be done but that will require some time, because I haven't found a single class (well, there is one ``NDDataBase``) which isn't depending on other changes.
- ``NDSlicingMixin`` could act as a standalone but will require that I at least delete one test (where the uncertainty is not an actual uncertainty but a np.ndarray).
- ``NDArithmeticMixin`` is not possible without ``NDUncertainty`` and not possible without the bugfix in ``NDData`` that sets the ``parent_nddata``.
- ``NDData`` and ``NDDataBase`` are standalone.

I could alter the slicing and arithmeticmixin slightly (actually I think that requires only few alterations) to make them standalone. Provided I don't have to disentangle ``NDArithmeticMixin`` and ``NDUncertainty``.

Thanke again for reviewing it and thanks for saying you'll review it again. I hope I didn't made too much mess with the last commits.

mwcraig commented:
@MSeifert04 

> I could alter the slicing and arithmeticmixin slightly (actually I think that requires only few alterations) to make them standalone. Provided I don't have to disentangle NDArithmeticMixin and NDUncertainty.

That would be great!Im thinking this PR (or the PRs it becomes) might be good candidates for a mini-sprint online once they are closer to ready? I was thinking small, the two of us and @wkerzendorf, and maybe one other if others are interested. Mostly the intent would be to arrange a few hours when we would all be online so we can talk through final changes/answer the little questions that come up/work through any roadblocks.

Thanks again for all of your work on this -- `NDArithmeticMixin` in particular was badly in need of refactoring, and I appreciate your patience with the code reviews!

MSeifert04 commented:
It seems like I can't root out some failures especially in the sphinx build (the doctests should pass, except one that failed for some reason in python2), I'll push as soon as I find the problems of the sphinx fails (if someone has an idea where they come from I could use some help).

@mwcraig About the splitting, to avoid duplication I'll wait for a few days before I start that but should I first post something on the dev-mailing list that more people might have a look on it before it is splitted?

I'd be very happy having a mini-sprint especially for the controversial or debatable changes. 

Let me know when you think the time is appropriate for splitting or organizing a dev-mail or a sprint. I think I did not get exactly in what order they should be. :-)

bsipocz commented on docs/nddata/index.rst:
```diff
@@ -156,6 +168,47 @@ Code that uses the arithemtic methods that used to be included in
+  ``meta`` attributes
+  like ``uncertainty``, ``wcs`` and ``mask``. But the Mixin was refactored in
+  a way that allows subclasses to customize slicing attributes in a different
+  way. Visit :ref:`nddata/subclassing` for more information.
```
one of the sphinx warning is that there is no such ``nddata/subclassing`` label, you should define it somewhere.

MSeifert04 replied:
Ah, that could be it, I thought sphinx would recognize ``files`` as ref. I'll look into it but it seems I will just create a ``subclassing`` label inside ``nddata/subclassing.rst``. Thanks!

bsipocz commented on docs/nddata/nddata.rst:
```diff
@@ -171,3 +204,8 @@ array like this::
+The same applies to extracting the data as Quantity::
+
+    >>> ndd_quantity = NDData([1, 2, 3, 4], unit="meter")
+    >>> quantity = u.Quantity(ndd_quantity, unit=ndd_quantity.unit)
```
this throws a TypeError for me. Also I wonder why one should add the nddata's unit as unit separately, shouldn't it be recognized by default?

```
UNEXPECTED EXCEPTION: TypeError(u'The value must be a valid Python or Numpy numeric type.',)
Traceback (most recent call last):

  File "/sw/lib/python2.7/doctest.py", line 1315, in __run
    compileflags, 1) in test.globs

  File "<doctest nddata.rst[34]>", line 1, in <module>

  File "astropy/units/quantity.py", line 264, in __new__
    raise TypeError("The value must be a valid Python or "

TypeError: The value must be a valid Python or Numpy numeric type.
```

MSeifert04 replied:
@bsipocz I forgot the ``.data`` while converting it. ``quantity = u.Quantity(ndd_quantity.unit, unit=ndd_quantity.unit)`` works.

Inside ``NDData`` the ``data`` and ``unit`` are seperated (that was one thing I did not change) so Quantity cannot recognize the unit by default.

bsipocz commented on astropy/nddata/nddata.py:
```diff
@@ -5,170 +5,239 @@
+    See also
+    --------
+
+    - `~astropy.nddata.mixins.NDIOMixin`
```
You should directly give the class/method/etc name here, and not as a link for that object. Also the namespaces must match to the ones in the api.

E.g.:

```
    See also
    --------
    astropy.nddata.NDIOMixin
    astropy.nddata.NDSlicingMixin
    astropy.nddata.NDArithmeticMixin
    astropy.nddata.NDDataArray
```

MSeifert04 replied:
Nice, I never thought that sphinx automatically parses the see also part as links. Thanks for the help!

bsipocz commented:
@MSeifert04 - My inlines should solve your sphinx issues, but I also got a separate fail for one of the doctests.

MSeifert04 commented:
@bsipocz Thanks for reviewing these errors. Ich changed them and pushed it. Seems like I'll have to read some of the points in the numpy/astropy/sphinx builder again I never thought even thought that these were the problem :+1:

mwcraig commented:
@MSeifert04 

>About the splitting, to avoid duplication I'll wait for a few days before I start that but should I first post something on the dev-mailing list that more people might have a look on it before it is splitted?

I think we'll get more reviewers after the split, honestly...this is a lot to review at once so splitting should make it easier for people to provide feedback.

> Let me know when you think the time is appropriate for splitting or organizing a dev-mail or a sprint. 

Let's do the split first, then pull together a mini-sprint/discussion. I'm happy to handle organizing the mini-sprint (unless you have a strong preference to do it yourself). We'll start with times the two of us and @wkerzendorf can make and go from there.

MSeifert04 commented:
@mwcraig Sorry, I forgot to answer: So you don't want to check the latest changes (as you said in https://github.com/astropy/astropy/pull/4234#issuecomment-149072655) and I should now start to split them?

mwcraig commented:
@MSeifert04 - could you please push a trivial change (e.g. white space ) to trigger a fresh Travis build just to make sure everything is working here before we try to split?

MSeifert04 commented:
@mwcraig I've pushed a change and it seems to be working.

mwcraig commented:
Excellent! Feel free to start splitting at your convenience...I'll be pretty available online Saturday and Sunday if you run into problems, need help or want advice (I'm UT -5).

Thanks again for all of your work on this!

MSeifert04 commented:
@mwcraig Sorry it took so long but I think I got the PR splitted:
#4270 , #4271 and #4272 (I didn't include the documentation-updates except if some doctests will fail). I had to work around some issues #4152 had to be included in the Mixin's PRs as well as some changes to work around ``UnknownUncertainty`` in the case of ``NDData`` but it should now work as planned.

mwcraig commented:
@MSeifert04  -- no worries, there was a lot to do! I'll take a look at them in a bit when the tests have completed.
